### ISSUE TITLE
Ready callbacks are called multiple times

### ISSUE THREAD

**Issue by rixlabs** [angular-google-chart/angular-google-chart#243]
I know you hate me :smile: 

Found out that the callback in agcOnReady  is called 2 times.

I also tried to create a directive similar to agcOnReady and I had the same problem, so is the internal callback that is called 2 times(abviously)

```
  link: function(scope, element, attrs, googleChartController){
                callback.$inject=['chartWrapper'];
                function callback(chartWrapper){
                    scope.$apply(function (){
                        scope.$eval(attrs.agcOnReady, {chartWrapper: chartWrapper});
                    });
                }
                googleChartController.registerWrapperListener('ready', callback, this);
            }
```

Debugging a little...
the registerWrapperListener is called by default in the _init and then for the agcOnReady.

The  $google.visualization.events.addListener is called 2 time with 2 different listener: _handleReady and callback(chartWrapper).

So is possible that the google library call the listener's handling_functions multiple times or is something with the ng-google-chart code?

p.s. the multi call is a problem for me cause I have to **append**  a node inside the DOM and I'm seeing it 2 times. As workarround I can check if that node exist but...

**Comment by Balrog30** [comment]
Listeners are registered throughout the module for various reasons, rather than trying to have one callback handle everything.

The Google Charts API triggers the ready event a lot. Basically, when you initialize the chart, and every time it draws. It's a signal that it is ready to take input, because it's done whatever you asked it to do. Multiple calls is desirable in Angular Google Chart's case, because it's how I'm keeping track of changes to the inner chart to re-register chart-specific events on the inner chart when the ChartWrapper destroys the chart and makes a new one.

It's also handy in a case like [this example](http://codepen.io/Balrog30/pen/yNZxKr) Where the ready event fires every time the chart redraws from a window resize, giving access the the layout interface to redraw the image overlays.

If you want to de-register an event listener, the registration function returns a function that you call to deregister. It's very much like the destroy functions returned by `$scope.$watch()`.

Since that deregistration function is available in the `agc-on-ready` directive, you could probably tack the de-registration function onto the locals for the $scope.$eval() call, to make it available as an injectable parameter in your listener function. That would be kind of a hacky way to do it, but it would work.

Your other options would be to keep track of the thing you want to do and ignore calls to your listener after it does it's thing the first time, or implement one-time listeners. I can do it or help set you up to do it if you like. The main reason I never got around to implementing them was that it was more of an internal API thing. I knew it could be useful, but unless someone goes and re-implements one-time versions of the event directives, there wasn't much point of making the one-time version of the listener registration.

Also, no one asked for it before.

**Comment by Balrog30** [comment]
By the way... I don't hate you. The project needs feedback and enquiries to keep it fresh. If there was no activity I would have nothing to do to improve Angular Google Chart, and people would call it "dead" and stop making new things with it.

**Comment by Balrog30** [comment]
```JavaScript
(function(){
    angular.module('googlechart')
        .directive('agcOnReady', onReadyDirective);
        
    function onReadyDirective(){
        return {
            restrict: 'A',
            scope: false,
            require: 'googleChart',
            link: function(scope, element, attrs, googleChartController){
                var destroy;
                callback.$inject=['chartWrapper'];
                function callback(chartWrapper){
                    scope.$apply(function (){
                        scope.$eval(attrs.agcOnReady, {chartWrapper: chartWrapper, destroy: destroy});
                    });
                }
                destory = googleChartController.registerWrapperListener('ready', callback, this);
            }
        };
    }
})();
```

```HTML
<div google-chart chart="chartObject" agc-on-ready="readyHandler(chartWrapper, destroy)"></div>
```

```JavaScript
$scope.readyHandler = function(chartWrapper, destroy) {
  destroy();
};
```

**Comment by rixlabs** [comment]
ok you reply a question a don't made,

there is no other way than modify the directive.

**Comment by rixlabs** [comment]
little problem here, you can't pass destroy in the locals cause you don't have it

**Comment by rixlabs** [comment]
Made a little directive that put a link to the image veriosn of the chart to teh bottom:
```
angular.module('googlechart')
	.directive('agcImageChart',agcImageChart);


function agcImageChart(){
	return {
		restrict: 'AE',
		scope: false,
		require: 'googleChart',
		link: function(scope, element, attrs, googleChartController){
			var destroy;
			callback.$inject = [ 'chartWrapper'];
			function callback(chartWrapper){
				tpmpl = '<a href="' + chartWrapper.getChart().getImageURI() + '" target=\"_blank\">Image version</a>'
				element.parent().append(tpmpl);
				destroy();
			}
			destroy = googleChartController.registerWrapperListener('ready', callback, this);
		}
	};
}
```


@Balrog30 I registered a Wrapper listener so the directive is indipendent. but I don't know if this is the best solution. And maybe I can pass the ID of the element where I want the link to the image

**Comment by Balrog30** [comment]
You stumbled across my least favourite feature of google charts: image uri. This is one of the features that I sort of looked at but not really when planning to have chart directives with different behaviour. The best way to handle this feature would be to have a chart directive that, instead of drawing, gets the imageUri and puts it in the source tag for an image. If I remember right about how I planned out the controller/service separation for the main directive, I think it might be possible. On a side note: how do you feel about TypeScript? I've been using it lately and creating method-compatible controllers for behaviour directives to use is a perfect use case for interfaces.

**Comment by rixlabs** [comment]
Hehe you don't like it? me neither but I'm working at a big statistics project and the stat expert need some static images for a print work(don't ask why he can't use print screen :smile: ). Sure you can use the image uri in a src in my case I used a target blank and let the browser do the work. 

About typescript I've made some test but then I started working on refactor my project with John Papa's style guide and abbandoned all the meta language thing(used also dart and coffee). I wanna try ECMA 6 with babel for a better protoyping. Typescript seems a good alternative too but I wanna try both(ECMA6 and Type) for a better comparison. btw is in my todo for next week(at work I'm migrating the GIS system to JS api and I have a time slot for studying) so I'll let you know.

**Comment by Balrog30** [comment]
Let me know if you want to schedule a video call or something for a walkthrough of the AGC internals. It would be nice to get some second opinions about how I set things up and to on-board a new contributor. When I started helping out, the original author didn't seem to have much time for the project, so I just kind of ran with it. I don't have a lot of time to write huge features, but I can certainly provide insight, answer questions, do maintenance, handle pull requests, and cut releases.
